src/core.rs

```diff
@@ -8,15 +8,60 @@
 
 use std::fmt;
 
+struct TreeWriter<'a, Item: ?Sized + fmt::Debug + 'a> {
```

Comment:
Dead code?

Speaking of, should we have a `ExpressionWriter` trait that `flatten` accepts as a reference to decouple the allocation policy from the `Predicate` trait?